backend/src/index.js

```diff
@@ -5,140 +5,68 @@
-
 const app = require( 'express' )();
 const { Server } = require( 'http' );
 const http = new Server( app );
 const io = require( 'socket.io' )( http );
+const dotenv = require( 'dotenv' );
+dotenv.config( { path: '.env' } );
 
-const db = new Map();
+const RedisDriver = require( './RedisDriver' );
+const SessionRepository = require( './SessionRepository' );
+const { UserHelper } = require( './helpers/UserHelper' );
 
-// Priorities are set from the lowest to the highest.
-const statePriorities = [ 'away', 'viewing', 'commenting', 'editing', 'merging' ];
+const driver = new RedisDriver( io );
+const repository = new SessionRepository( driver );
 
 io.on( 'connection', socket => {
-	socket.session = {};
-
 	const timestamp = new Date();
 
-	socket.on( 'setUser', ( message, reply ) => {
-		const issueKey = `${ message.repoName }:${ message.issueId }`;
-		if ( !db.has( issueKey ) ) {
-			db.set( issueKey, new Map() );
-		}
-
-		if ( !socket.session.issues ) {
-			socket.session.issues = [];
-		}
-
-		if ( !socket.session.issues.includes( issueKey ) ) {
-			socket.session.issues.push( issueKey );
-		}
-
-		const issueUsers = db.get( issueKey );
-
-		if ( !issueUsers.has( message.user.login ) ) {
-			issueUsers.set( message.user.login, { sockets: [], tabs: new Map(), joinedAt: timestamp } );
-		}
-
-		const issueUser = issueUsers.get( message.user.login );
+	socket.on( 'setUser', async ( message, reply ) => {
+		const issueKey = createIssueKey( message.repoName, message.pageType, message.issueId );
 
-		if ( !issueUser.sockets.includes( socket.id ) ) {
-			socket.session.login = message.user.login;
+		const issueSession = { ...message.user, joinedAt: timestamp };
 
-			socket.join( issueKey );
+		await repository.createOrUpdate( issueKey, socket.id, issueSession );
 
-			issueUser.sockets.push( socket.id );
+		if ( !socket.issueKey ) {
+			socket.issueKey = issueKey;
 		}
 
-		issueUser.tabs.set( socket.id, message.user.state );
+		const issueSessions = await repository.getAll( issueKey );
 
-		issueUser.user = message.user;
+		socket.join( issueKey );
 
-		const users = getUsers( issueUsers );
-
-		socket.broadcast.to( issueKey ).emit( 'refresh', users );
+		const users = UserHelper.getUserListFromSessions( issueSessions );
 
 		reply( null, users );
-	} );
 
-	socket.on( 'removeUser', message => {
-		const issueKey = `${ message.repoName }:${ message.issueId }`;
-
-		removeUser( issueKey );
+		socket.broadcast.to( issueKey ).emit( 'refresh', users );
 	} );
 
-	socket.on( 'disconnect', () => {
-		for ( const issueKey of socket.session.issues || [] ) {
-			removeUser( issueKey );
-		}
-
-		delete socket.issues;
-	} );
+	socket.on( 'disconnect', async () => {
+		await repository.deleteOne( socket.id );
 
-	function removeUser( issueKey ) {
-		const issueUsers = db.get( issueKey );
+		const issueSessions = await repository.getAll( socket.issueKey );
 
-		if ( !issueUsers ) {
+		if ( !issueSessions[ socket.id ] ) {
```

Comment:
It isn't necessary. And we should in both cases call the `socket.leave( socket.issueKey );` I guess.

Reply:
Sure, done.